checker/src/visitors.rs

```diff
@@ -497,44 +493,19 @@ impl<'analysis, 'compilation, 'tcx, E> MirVisitor<'analysis, 'compilation, 'tcx,
         actual_argument_types: &[Ty<'tcx>],
         generic_args: Option<SubstsRef<'tcx>>,
     ) -> Option<Summary> {
-        let mut result = None;
+        debug!("devirt def_id {:?}", def_id);
+        debug!("devirt generic_args {:?}", generic_args);
         if !actual_args.is_empty() && !self.tcx.is_closure(self.def_id) {
-            if let Expression::Variable { path, .. } = &actual_args[0].1.expression {
-                let mut self_ty = self.get_path_rustc_type(path);
-                if let TyKind::Ref(_, ty, _) = self_ty.kind {
-                    self_ty = ty;
-                }
-                match self_ty.kind {
-                    TyKind::Adt(..) => (),
-                    _ => {
-                        return None;
-                    }
-                }
-                let mut might_resolve = true;
-                let substs = InternalSubsts::for_item(self.tcx, def_id, |param_def, _| {
-                    match param_def.kind {
-                        rustc::ty::GenericParamDefKind::Lifetime
-                        | rustc::ty::GenericParamDefKind::Const => {}
-                        rustc::ty::GenericParamDefKind::Type { .. } => {
-                            if param_def.index == 0 {
-                                return self_ty.into();
-                            } else {
-                                might_resolve = false;
-                            }
-                        }
-                    }
-                    self.tcx.mk_param_from_def(param_def)
-                });
-                if might_resolve {
+            if let Some(gen_args) = generic_args {
+                if utils::are_concrete(&gen_args) {
```

Comment:
This is a bit sad, but without it the rust compiler panics in all sorts of ways that are hard to explain. I spent two days trying to figure out how to do better, but this is what I can get to work right now.